src/Streamly/Internal/Data/Parser/Tee.hs

```diff
@@ -0,0 +1,297 @@
+{-# LANGUAGE CPP                       #-}
+
+-- Somehow uni-pattern match using "let" produces better optimized code
+-- compared to using cases and using explicit "error" messages in unreachable
+-- cases.  There seem to be no way to silence individual warning so we use this
+-- global option for the file.
+--
+-- Disabling this has the potential to mask off some legit warnings, therefore
+-- we have segregated code that uses uni-pattern matches in this file.
+#if __GLASGOW_HASKELL__ >= 800
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+#endif
+
+#include "inline.hs"
+
+-- |
+-- Module      : Streamly.Internal.Data.Parser.Tee
+-- Copyright   : (c) 2020 Composewell Technologies
+-- License     : BSD3
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- Parsers.
+
+module Streamly.Internal.Data.Parser.Tee
+    (
+    -- Parallel parsers
+      teeWith
+    , teeWithFst
+    , teeWithMin
+    )
+where
+
+import Control.Exception (assert)
+import Prelude
+       hiding (any, all, takeWhile)
+
+import Fusion.Plugin.Types (Fuse(..))
+import Streamly.Internal.Data.Parser.Types (Parser(..), Step(..))
+
+-------------------------------------------------------------------------------
+-- Distributing Parsers
+-------------------------------------------------------------------------------
+
+data StepState s a = StepState s | StepResult a
+
+-- | State of the pair of parsers in a tee composition
+-- Note: strictness annotation is important for fusing the constructors
+{-# ANN type TeeState Fuse #-}
+data TeeState sL sR x a b =
+    TeePair !([x], StepState sL a, [x], [x]) !([x], StepState sR b, [x], [x])
```

Comment:
I think it'd be helpful if you can document what all the arguments here are supposed to be, would be useful while visiting this function in the future.

Reply:
yes.